File: morning_exercises/day31.cpp
#include <string>
#include <unordered_map>
#include <iostream>

std::string longestSubstring(std::string input) {
    int left = 0; int right = 0; int maxVal = 0; int maxLeft=0; int maxRight=0;
    std::unordered_map<char, int> map;

    while (left<input.size() && right<input.size()) {
        if (input.empty()) return "";

        if (map[input[right]]>0) {
            map[input[left]]--;
            left++;
        }
        else {
            if ((right-left)>=maxVal) {
                maxVal=right-left;
                maxLeft = left;
                maxRight = right;
            }
            map[input[right]]++;
            right++;
        }
    }
    return input.substr(maxLeft, maxVal+1);
}

int main() {
    std::cout << longestSubstring("abcabcbb") << std::endl; // "abc"
    std::cout << longestSubstring("bbbbb") << std::endl;    // "b"
    std::cout << longestSubstring("pwwkew") << std::endl;   // "wke"
    std::cout << longestSubstring("abcdef") << std::endl;   // "abcdef"
    std::cout << longestSubstring("aab") << std::endl;      // "ab"
    std::cout << longestSubstring("dvdf") << std::endl;     // "vdf"
    std::cout << longestSubstring("a") << std::endl;        // "a"
}

File: morning_exercises/day31.ts
function longestSubstring(input: string): string {
    let left = 0, right = 0, maxVal = 0, maxLeft = 0, maxRight = 0;
    const map = new Map<string, number>();

    while (left<input.length && right<input.length) {
        if ((map.get(input[right]) ?? 0)>0) {
            map.set(input[left], map.get(input[left])! - 1)
            left++;
        }
        else {
            if ((right-left)>=maxVal) {
                maxVal=right-left;
                maxLeft=left; maxRight = right;
            } 
            map.set(input[right], (map.get(input[right]) ?? 0)!+1)
            right++;
        }
    }
    return input.substring(maxLeft,maxRight+1);
}

console.log(longestSubstring("abcabcbb")); // "abc"
console.log(longestSubstring("bbbbb"));    // "b"
console.log(longestSubstring("pwwkew"));   // "wke"
console.log(longestSubstring("abcdef"));   // "abcdef"
console.log(longestSubstring("aab"));      // "ab"
console.log(longestSubstring("dvdf"));     // "vdf"
console.log(longestSubstring("a"));        // "a"
console.log(longestSubstring(""));         // ""